src/app/components/common/node-editor/connector.ts

```diff
@@ -14,8 +17,26 @@ export class Connector<S extends ClassicScheme, K extends any[]> extends Bidirec
           return;
         }
 
+        const connectionExist = context.editor.getConnections().find(conn => (conn.source == initial.nodeId && conn.target == socket.nodeId));
+        if (connectionExist) {
+          return;
+        }
+
         const fromNode = context.editor.getNode(initial.nodeId);
         const toNode = context.editor.getNode(socket.nodeId);
+        const isNorthNode = context.editor.getNodes().find(node => node.label == 'North');
+
+        // Invalid connection check
+        if (
+          (fromNode instanceof Storage && toNode instanceof South) ||
+          (fromNode instanceof North && toNode instanceof Storage) ||
+          (fromNode instanceof Filter && toNode instanceof South) ||
+          (fromNode instanceof North && toNode instanceof Filter) ||
+          (isNorthNode && fromNode instanceof Filter && toNode instanceof Storage)
```

Comment:
I would rather have a validConnections array of objects with src, dst and if does not have any return with console log